Tribler/Policies/BoostingManager.py

```diff
@@ -0,0 +1,1105 @@
+            # not using Session.start_download because we need to specify pstate
+            assert self.session.get_libtorrent()
+
+            tobj['download'] = self.session.lm.add(tobj['metainfo'], dscfg, pstate=tobj.get('pstate', None),
+                                                   hidden=True, share_mode=not preload, checkpoint_disabled=True)
+            tobj['download'].set_priority(tobj.get('prio', 1))
+
+        self.session.lm.threadpool.add_task_in_thread(do_start, 0)
+
+    def stop_download(self, torrent):
+        """
+        Stopping torrent that currently downloading
+        """
+        def do_stop():
+            """
+            The actual function to stop torrent downloading
+            :return:
+            """
+            ihash = lt.big_number(torrent["metainfo"].get_infohash())
+            self._logger.info("Stopping %s", str(ihash))
+            download = torrent.pop('download', False)
+            lt_torrent = self.session.lm.ltmgr.get_session().find_torrent(ihash)
+            if download and lt_torrent.is_valid():
+                self._logger.debug("Writing resume data")
+                torrent['pstate'] = {'engineresumedata': download.write_resume_data()}
+                self.session.remove_download(download)
+
+        self.session.lm.threadpool.add_task_in_thread(do_stop, 0)
+
+    def _select_torrent(self):
+        """
+        Function to select which torrent in the torrent list will be downloaded in the
+        next iteration. It depends on the source and applied policy
+        """
+        torrents = {}
+        for infohash, torrent in self.torrents.iteritems():
+            # we prioritize archive source
+            if torrent.get('preload', False):
+                if 'download' not in torrent:
+                    self.start_download(torrent)
+                elif torrent['download'].get_status() == DLSTATUS_SEEDING:
+                    self.stop_download(torrent)
+            elif not torrent.get('is_duplicate', False):
+                if torrent.get('enabled', True):
+                    torrents[infohash] = torrent
+
+        if self.policy is not None and torrents:
+            # Determine which torrent to start and which to stop.
+            torrents_start, torrents_stop = self.policy.apply(
+                torrents, self.max_torrents_active)
+            for torrent in torrents_stop:
+                self.stop_download(torrent)
+            for torrent in torrents_start:
+                self.start_download(torrent)
+
+            self._logger.info("Selecting from %s torrents %s start download", len(torrents), len(torrents_start))
+
+        self.session.lm.threadpool.add_task(self._select_torrent, self.swarm_interval)
+
+    def load_config(self):
+        config = ConfigParser.ConfigParser()
+        config.read(self.config_file)
+        validate_source = lambda s: unhexlify(s) if len(s) == 40 and not s.startswith("http") else s
+        for k, val in config.items(__name__):
```

Comment:
There are (I think too) many branches in here. Is it somehow possible to clean up the branches and reduce the complexity? Maybe you can do something with dictionaries, see [here](http://stackoverflow.com/questions/11479816/what-is-the-python-equivalent-for-a-case-switch-statement) for some inspiration.